File: creator-node/src/monitors/monitors.js
const { getDatabaseSize, getDatabaseConnections, getDatabaseConnectionInfo, getDatabaseLiveness } = require('./database')
const {
  getTotalMemory,
  getUsedMemory,
  getNodeProcessMemoryUsage,
  getUsedTCPMemory
} = require('./memory')
const {
  getStoragePathSize,
  getStoragePathUsed,
  getFilesystemSize,
  getFilesystemUsed,
  getMaxFileDescriptors,
  getAllocatedFileDescriptors
} = require('./filesystem')
const {
  getReceivedBytesPerSec,
  getTransferredBytesPerSec
} = require('./network')
const {
  getRedisNumKeys,
  getRedisUsedMemory,
  getRedisTotalMemory
} = require('./redis')
const {
  getIPFSReadWriteStatus
} = require('./ipfs')
const {
  get30DayRollingSyncSuccessCount,
  get30DayRollingSyncFailCount,
  getDailySyncSuccessCount,
  getDailySyncFailCount,
  getLatestSyncSuccessTimestamp,
  getLatestSyncFailTimestamp
} = require('./syncHistory')
const redis = require('../redis')

// Prefix used to key each monitored value in redis
const MONITORING_REDIS_PREFIX = 'monitoring'

/**
 * List of all monitors to run, containing:
 *  @param {string} name A unique name for the metric (for caching in redis)
 *  @param {function} func The actual work to compute a value. The return value is what is cached.
 *  @param {number?} ttl TTL in seconds for how long a cached value is good for.
 *    Since the job runs on a cron, the min TTL a metric can be refreshed is 60s.
 *    If a TTL isn't provided, the metric is refreshed every 60s.
 *  @param {string?} type Optional type that the value should be parsed to (default is string)
 *    Options are bool, int, float, string, json
 */

const DATABASE_LIVENESS = {
  name: 'databaseLiveness',
  func: getDatabaseLiveness,
  type: 'bool'
}
const DATABASE_SIZE = {
  name: 'databaseSize',
  func: getDatabaseSize,
  ttl: 60 * 2,
  type: 'int'
}
const DATABASE_CONNECTIONS = {
  name: 'databaseConnections',
  func: getDatabaseConnections,
  type: 'int'
}
const DATABASE_CONNECTION_INFO = {
  name: 'databaseConnectionInfo',
  func: getDatabaseConnectionInfo,
  type: 'json'
}

const TOTAL_MEMORY = {
  name: 'totalMemory',
  func: getTotalMemory,
  type: 'int'
}
const USED_MEMORY = {
  name: 'usedMemory',
  func: getUsedMemory,
  type: 'int'
}
const NODE_PROCESS_MEMORY_USAGE = {
  name: 'nodeProcessMemoryUsage',
  func: getNodeProcessMemoryUsage,
  type: 'json'
}

const USED_TCP_MEMORY = {
  name: 'usedTCPMemory',
  func: getUsedTCPMemory,
  type: 'int'
}

const STORAGE_PATH_SIZE = {
  name: 'storagePathSize',
  func: getStoragePathSize,
  ttl: 60 * 5,
  type: 'int'
}
const STORAGE_PATH_USED = {
  name: 'storagePathUsed',
  func: getStoragePathUsed,
  ttl: 60 * 5,
  type: 'int'
}
const FILESYSTEM_SIZE = {
  name: 'filesystemSize',
  func: getFilesystemSize,
  ttl: 60 * 5,
  type: 'int'
}
const FILESYSTEM_USED = {
  name: 'filesystemUsed',
  func: getFilesystemUsed,
  ttl: 60 * 5,
  type: 'int'
}
const MAX_FILE_DESCRIPTORS = {
  name: 'maxFileDescriptors',
  func: getMaxFileDescriptors,
  ttl: 60 * 5,
  type: 'int'
}
const ALLOCATED_FILE_DESCRIPTORS = {
  name: 'allocatedFileDescriptors',
  func: getAllocatedFileDescriptors,
  ttl: 60 * 5,
  type: 'int'
}

const RECEIVED_BYTES_PER_SEC = {
  name: 'receivedBytesPerSec',
  func: getReceivedBytesPerSec,
  type: 'float'
}
const TRANSFERRED_BYTES_PER_SEC = {
  name: 'transferredBytesPerSec',
  func: getTransferredBytesPerSec,
  type: 'float'
}

const REDIS_NUM_KEYS = {
  name: 'redisNumKeys',
  func: getRedisNumKeys,
  ttl: 60 * 5,
  type: 'int'
}
const REDIS_USED_MEMORY = {
  name: 'redisUsedMemory',
  func: getRedisUsedMemory,
  ttl: 60 * 5,
  type: 'int'
}
const REDIS_TOTAL_MEMORY = {
  name: 'redisTotalMemory',
  func: getRedisTotalMemory,
  ttl: 60 * 5,
  type: 'int'
}

const IPFS_READ_WRITE_STATUS = {
  name: 'IPFSReadWriteStatus',
  func: getIPFSReadWriteStatus,
  ttl: 60 * 5,
  type: 'json'
}

// The rolling window count of successful syncs
// The window is for 30 days
// Keep the rolling window count ttl to 1 hour (refreshes every hour)
const THIRTY_DAY_ROLLING_SYNC_SUCCESS_COUNT = {
  name: 'rollingSyncSuccessCount',
  func: get30DayRollingSyncSuccessCount,
  ttl: 60 /* mins */ * 60 /* s */,
  type: 'int'
}

// The rolling window count of failed syncs
// The window is for 30 days
// Keep the rolling window count ttl to 1 hour (refreshes every hour)
const THIRTY_DAY_ROLLING_SYNC_FAIL_COUNT = {
  name: 'rollingSyncFailCount',
  func: get30DayRollingSyncFailCount,
  ttl: 60 /* mins */ * 60 /* s */,
  type: 'int'
}

// The daily count of successful syncs
// Set ttl to every 5 minutes
const DAILY_SYNC_SUCCESS_COUNT = {
  name: 'dailySyncSuccessCount',
  func: getDailySyncSuccessCount,
  ttl: 5 /* mins */ * 60 /* s */,
  type: 'int'
}

// The daily count of successful syncs
// Set ttl freshness to every 5 minutes
const DAILY_SYNC_FAIL_COUNT = {
  name: 'dailySyncFailCount',
  func: getDailySyncFailCount,
  ttl: 5 /* mins */ * 60 /* s */,
  type: 'int'
}

// The timestamp of the latest succesful sync
// Set ttl freshness to every 5 minutes
const LATEST_SYNC_SUCCESS_TIMESTAMP = {
  name: 'latestSyncSuccessTimestamp',
  func: getLatestSyncSuccessTimestamp,
  ttl: 5 /* mins */ * 60 /* s */,
  type: 'string'
}

// The timestamp of the latest failed sync
// Set ttl freshness to every 5 minutes
const LATEST_SYNC_FAIL_TIMESTAMP = {
  name: 'latestSyncFailTimestamp',
  func: getLatestSyncFailTimestamp,
  ttl: 5 /* mins */ * 60 /* s */,
  type: 'string'
}

const MONITORS = {
  DATABASE_LIVENESS,
  DATABASE_SIZE,
  DATABASE_CONNECTIONS,
  DATABASE_CONNECTION_INFO,
  TOTAL_MEMORY,
  USED_MEMORY,
  NODE_PROCESS_MEMORY_USAGE,
  USED_TCP_MEMORY,
  STORAGE_PATH_SIZE,
  STORAGE_PATH_USED,
  FILESYSTEM_SIZE,
  FILESYSTEM_USED,
  MAX_FILE_DESCRIPTORS,
  ALLOCATED_FILE_DESCRIPTORS,
  RECEIVED_BYTES_PER_SEC,
  TRANSFERRED_BYTES_PER_SEC,
  REDIS_NUM_KEYS,
  REDIS_USED_MEMORY,
  REDIS_TOTAL_MEMORY,
  IPFS_READ_WRITE_STATUS,
  THIRTY_DAY_ROLLING_SYNC_SUCCESS_COUNT,
  THIRTY_DAY_ROLLING_SYNC_FAIL_COUNT,
  DAILY_SYNC_SUCCESS_COUNT,
  DAILY_SYNC_FAIL_COUNT,
  LATEST_SYNC_SUCCESS_TIMESTAMP,
  LATEST_SYNC_FAIL_TIMESTAMP
}

const getMonitorRedisKey = (monitor) => `${MONITORING_REDIS_PREFIX}:${monitor.name}`

/**
 * Parses a string value into the corresponding type
 * @param {Object} monitor
 * @param {string} value
 */
const parseValue = (monitor, value) => {
  try {
    if (monitor.type) {
      switch (monitor.type) {
        case 'bool':
          return value === 'true'
        case 'int':
          return parseInt(value)
        case 'float':
          return parseFloat(value)
        case 'json':
          return JSON.parse(value)
        case 'string':
          return value
        default:
          return value
      }
    }
    return value
  } catch (e) {
    return value
  }
}

/**
 * Gets monitor values
 * @param {Array<Object>} monitors the monitor, containing name, func, ttl, and type
 */
const getMonitors = async (monitors) => {
  const pipeline = redis.pipeline()
  monitors.forEach(monitor => {
    const key = getMonitorRedisKey(monitor)
    pipeline.get(key)
  })
  return pipeline
    .exec()
    // Pull the value off of the result
    .then((result) => result.map((r, i) => parseValue(monitors[i], r[1])))
}

module.exports = {
  MONITORS,
  getMonitorRedisKey,
  getMonitors
}

File: discovery-provider/src/monitors/monitors.py
import json
from src.monitors import monitor_names
from src.monitors.database import (
    get_database_connection_info,
    get_database_connections,
    get_database_liveness,
    get_database_size,
    get_database_index_count,
    get_database_index_info,
    get_table_size_info,
    get_frequent_queries,
    get_slow_queries,
)
from src.monitors.memory import get_total_memory, get_used_memory
from src.monitors.filesystem import get_filesystem_size, get_filesystem_used
from src.monitors.network import (
    get_received_bytes_per_sec,
    get_transferred_bytes_per_sec,
)
from src.monitors.redis import (
    get_redis_num_keys,
    get_redis_used_memory,
    get_redis_total_memory,
)
from src.utils import redis_connection

redis = redis_connection.get_redis()


MONITORING_REDIS_PREFIX = "monitoring"

"""
    List of all monitors to run, containing
    Args:
    name: string The name of the metric (used to key into redis )
    func: function The function to compute the metric. Takes in kwargs
        db = The singleton db instance
        redis = The singleton redis instance
    ttl: int (optional) TTL in seconds for how long a cached value is good for, min 60s
    type: int (optional) type that the value should be parsed to (default is str)
        Options are bool, int, float, str, json
"""

DATABASE_LIVENESS = {
    monitor_names.name: monitor_names.database_liveness,
    monitor_names.func: get_database_liveness,
    monitor_names.type: "bool",
}
DATABASE_SIZE = {
    monitor_names.name: monitor_names.database_size,
    monitor_names.func: get_database_size,
    monitor_names.ttl: 60 * 2,
    monitor_names.type: "int",
}
DATABASE_CONNECTIONS = {
    monitor_names.name: monitor_names.database_connections,
    monitor_names.func: get_database_connections,
    monitor_names.type: "int",
}
DATABASE_CONNECTION_INFO = {
    monitor_names.name: monitor_names.database_connection_info,
    monitor_names.func: get_database_connection_info,
    monitor_names.type: "json",
}

DATABASE_INDEX_COUNT = {
    monitor_names.name: monitor_names.database_index_count,
    monitor_names.func: get_database_index_count,
    monitor_names.ttl: 60 * 60 * 6,  # six hours
    monitor_names.type: "int",
}

DATABASE_INDEX_INFO = {
    monitor_names.name: monitor_names.database_index_info,
    monitor_names.func: get_database_index_info,
    monitor_names.ttl: 60 * 60 * 6,  # six hours
    monitor_names.type: "json",
}

TABLE_SIZE_INFO = {
    monitor_names.name: monitor_names.table_size_info,
    monitor_names.func: get_table_size_info,
    monitor_names.type: "json",
}

FREQUENT_QUERIES = {
    monitor_names.name: monitor_names.frequent_queries,
    monitor_names.func: get_frequent_queries,
    monitor_names.type: "json",
}

SLOW_QUERIES = {
    monitor_names.name: monitor_names.slow_queries,
    monitor_names.func: get_slow_queries,
    monitor_names.type: "json",
}

TOTAL_MEMORY = {
    monitor_names.name: monitor_names.total_memory,
    monitor_names.func: get_total_memory,
    monitor_names.ttl: 60 * 2,
    monitor_names.type: "int",
}
USED_MEMORY = {
    monitor_names.name: monitor_names.used_memory,
    monitor_names.func: get_used_memory,
    monitor_names.ttl: 60 * 2,
    monitor_names.type: "int",
}

FILESYSTEM_SIZE = {
    monitor_names.name: monitor_names.filesystem_size,
    monitor_names.func: get_filesystem_size,
    monitor_names.ttl: 60 * 5,
    monitor_names.type: "int",
}
FILESYSTEM_USED = {
    monitor_names.name: monitor_names.filesystem_used,
    monitor_names.func: get_filesystem_used,
    monitor_names.ttl: 60 * 5,
    monitor_names.type: "int",
}

RECEIVED_BYTES_PER_SEC = {
    monitor_names.name: monitor_names.received_bytes_per_sec,
    monitor_names.func: get_received_bytes_per_sec,
    monitor_names.type: "float",
}

TRANSFERRED_BYTES_PER_SEC = {
    monitor_names.name: monitor_names.transferred_bytes_per_sec,
    monitor_names.func: get_transferred_bytes_per_sec,
    monitor_names.type: "float",
}

REDIS_NUM_KEYS = {
    monitor_names.name: monitor_names.redis_num_keys,
    monitor_names.func: get_redis_num_keys,
    monitor_names.ttl: 60 * 5,
    monitor_names.type: "int",
}
REDIS_USED_MEMORY = {
    monitor_names.name: monitor_names.redis_used_memory,
    monitor_names.func: get_redis_used_memory,
    monitor_names.ttl: 60 * 5,
    monitor_names.type: "int",
}
REDIS_TOTAL_MEMORY = {
    monitor_names.name: monitor_names.redis_total_memory,
    monitor_names.func: get_redis_total_memory,
    monitor_names.ttl: 60 * 5,
    monitor_names.type: "int",
}

MONITORS = {
    monitor_names.database_liveness: DATABASE_LIVENESS,
    monitor_names.database_size: DATABASE_SIZE,
    monitor_names.database_connections: DATABASE_CONNECTIONS,
    monitor_names.database_connection_info: DATABASE_CONNECTION_INFO,
    monitor_names.database_index_count: DATABASE_INDEX_COUNT,
    monitor_names.database_index_info: DATABASE_INDEX_INFO,
    monitor_names.table_size_info: TABLE_SIZE_INFO,
    monitor_names.frequent_queries: FREQUENT_QUERIES,
    monitor_names.slow_queries: SLOW_QUERIES,
    monitor_names.total_memory: TOTAL_MEMORY,
    monitor_names.used_memory: USED_MEMORY,
    monitor_names.filesystem_size: FILESYSTEM_SIZE,
    monitor_names.filesystem_used: FILESYSTEM_USED,
    monitor_names.received_bytes_per_sec: RECEIVED_BYTES_PER_SEC,
    monitor_names.transferred_bytes_per_sec: TRANSFERRED_BYTES_PER_SEC,
    monitor_names.redis_num_keys: REDIS_NUM_KEYS,
    monitor_names.redis_used_memory: REDIS_USED_MEMORY,
    monitor_names.redis_total_memory: REDIS_TOTAL_MEMORY,
}


def get_monitor_redis_key(monitor):
    return f"{MONITORING_REDIS_PREFIX}:{monitor[monitor_names.name]}"


# pylint: disable=R0911
def parse_value(monitor, value):
    """
    Parses a string value into the corresponding type

    Args:
        monitor: dict The monitor dictionary qwith name, func, ttl, and type
        value: string The value to parse
    """
    try:
        if str(value) == "None":
            return None
        # pylint: disable=R1705
        if monitor[monitor_names.type] == "bool":
            return value == "True"
        elif monitor[monitor_names.type] == "int":
            return int(value)
        elif monitor[monitor_names.type] == "float":
            return float(value)
        elif monitor[monitor_names.type] == "json":
            return json.loads(value)
        else:  # string
            return str(value)
    except Exception:
        return str(value)


def get_monitors(monitors):
    """
    Gets monitor values

    Args:
        monitors: list(dict) The list of monitors to get values of
    """
    pipe = redis.pipeline()
    for monitor in monitors:
        key = get_monitor_redis_key(monitor)
        pipe.get(key)
    ret = {}
    results = pipe.execute()
    for i, result in enumerate(results):
        ret[monitors[i][monitor_names.name]] = parse_value(monitors[i], result)
    return ret
